glommio/src/executor/latch.rs

```diff
@@ -0,0 +1,227 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the MIT/Apache-2.0 License, at your convenience
+//
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2020 Datadog, Inc.
+//
+//! Similar to a [`std::sync::Barrier`] but provides [`Latch::cancel`] which a
+//! failed thread can use to cancel the `Latch`.
+//! [`Latch::wait`] and [`Latch::arrive_and_wait`] return a [`LatchState`] to
+//! determine whether the state is `Ready` or `Canceled`.
+//!
+//! The implementation is intended for multi-threaded rather than task local
+//! use.
+
+use std::sync::{
+    atomic::{AtomicUsize, Ordering},
+    Arc,
+    Condvar,
+    Mutex,
+};
+
+#[derive(Clone, Debug)]
+pub(crate) struct Latch {
```

Comment:
I am still reading through, so take this with a grain of salt: but I suspect this would be a useful primitive to have in `glommio::sync`! It's essentially the async version of a barrier.

I don't anticipate widespread use (the same way barriers aren't used that widely either), but I can definitely see them being useful in some circumstances.

Won't affect the decision of what we do with this PR, but if you agree consider moving it later as a `pub` interface.

Reply:
Ok, I already see that it has mutexes internally which make it harder to use in generic code.
We could have a version of that over shared channels, but that's for later.

It's still a good idea, and I'll consider opening a new issue for that.

Thanks!